doc/manual/metaprogramming.rst

```diff
@@ -893,37 +893,46 @@ the arguments are known, but the function is not yet compiled.
 Instead of performing some calculation or action, a generated function
 declaration returns a quoted expression which then forms the body for the
 method corresponding to the types of the arguments. When called, the body
-expression is compiled (or fetched from a cache, on subsequent calls) and
-only the returned expression - not the code that generated it - is evaluated.
+expression is first evaluated and compiled,
+then the returned expression is compiled and run.
+To make this efficient, the result is often cached.
+And to make this inferable, only a limited subset of the language is usable.
 Thus, generated functions provide a flexible framework to move work from
-run-time to compile-time.
+run-time to compile-time, at the expense of greater restrictions on the allowable constructs.
 
-When defining generated functions, there are three main differences to
+When defining generated functions, there are four main differences to
 ordinary functions:
 
 1. You annotate the function declaration with the ``@generated`` macro.
    This adds some information to the AST that lets the compiler know that
    this is a generated function.
 
 2. In the body of the generated function you only have access to the
-   *types* of the arguments, not their values.
+   *types* of the arguments -- not their values -- and any function that
+   was defined *before* the definition of generated function.
 
 3. Instead of calculating something or performing some action, you return
    a *quoted expression* which, when evaluated, does what you want.
 
+4. The function must not have any side-effects or examine any non-constant
+   global state (including, for example, IO, locks, or non-local dictionaries).
+   In other words, they must be completely pure.
+   Due to an implementation limitation,
+   this also means that they currently cannot define a closure or untyped generator.
+
 It's easiest to illustrate this with an example. We can declare a generated
 function ``foo`` as
 
 .. doctest::
 
     julia> @generated function foo(x)
-               println(x)
-               return :(x*x)
+               Core.println(x)
```

Comment:
`println` isn't OK, it's just a convenient function for showing how they work (and along the way demonstrating why it isn't OK)